{
 "metadata": {
  "name": "",
  "signature": "sha256:9b0e71ec035866dd9130db3c2d210741d9aed7a0e5c298a9f38bd4e0a2f4f2b3"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from __future__ import print_function"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from curation_common import *"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pandas\n",
      "import requests\n",
      "import gcat\n",
      "import jsonschema\n",
      "import types\n",
      "from IPython import display\n",
      "import RDF\n",
      "\n",
      "from htsworkflow.util import hashfile"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "ImportError",
       "evalue": "cannot import name hashlib",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
        "\u001b[0;32m<ipython-input-21-30c6962c4e30>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mRDF\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0mhtsworkflow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mutil\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mhashlib\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[0;31mImportError\u001b[0m: cannot import name hashlib"
       ]
      }
     ],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dry_run = True"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "server = ENCODED('www.encodedcc.org')\n",
      "server.load_netrc()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "schemas = {\n",
      " '/experiments/': server.get_json('/profiles/experiment.json'),\n",
      " '/libraries/': server.get_json('/profiles/library.json'),\n",
      " '/replicates/': server.get_json('/profiles/replicate.json'),\n",
      " '/biosamples/': server.get_json('/profiles/biosample.json'),\n",
      "}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def post(collection, data, dry_run):\n",
      "    try:\n",
      "        jsonschema.validate(data, schemas[collection])\n",
      "        if not dry_run:\n",
      "            response =  server.post_json(collection, data)\n",
      "            obj = response['@graph'][0]\n",
      "            return obj['@id']\n",
      "    except requests.HTTPError as e:\n",
      "        print('exception: {}'.format(e))\n",
      "        display.display_pretty(data)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def parse_sheet_array_type(value):\n",
      "    return value.split(', ')\n",
      "\n",
      "def parse_sheet_integer_type(value):\n",
      "    return int(value)\n",
      "\n",
      "def parse_sheet_boolean_type(value):\n",
      "    return bool(value)\n",
      "\n",
      "def parse_sheet_timestamp_type(value):\n",
      "    return value.strftime('%Y-%m-%d')\n",
      "\n",
      "def parse_sheet_string_type(value):\n",
      "    return unicode(value)\n",
      "\n",
      "parsers = {\n",
      "  'array': parse_sheet_array_type,\n",
      "  'boolean': parse_sheet_boolean_type,\n",
      "  'integer': parse_sheet_integer_type,\n",
      "  'date': parse_sheet_timestamp_type,\n",
      "}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def parse_sheet_cell(header, value):\n",
      "    header = header.split(':')\n",
      "    parse_function = parse_sheet_string_type\n",
      "    if len(header) > 1:\n",
      "        if header[1] == 'skip':\n",
      "            return None, None\n",
      "        parse_function = parsers[header[1]]\n",
      "    return header[0], parse_function(value)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "book_name = 'ENCODE Submission for Y3Q1'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def post_biosamples(book_name, dry_run=True):\n",
      "    book = gcat.get_file(book_name, fmt='pandas_excel')\n",
      "    biosamples = book.parse('Biosamples', header=0)\n",
      "    for i in biosamples.index:\n",
      "        row = biosamples.ix[i]\n",
      "        columns = biosamples.columns\n",
      "        accession = row['biosample_accession']\n",
      "        if pandas.isnull(accession):\n",
      "            new_object = {}\n",
      "            for k in columns:\n",
      "                if pandas.notnull(row[k]):\n",
      "                    name, value = parse_sheet_cell(k, row[k])\n",
      "                    if name is None:\n",
      "                        continue\n",
      "                    if name == 'starting_amount_units' and value == 'pg':\n",
      "                        del new_object['starting_amount']\n",
      "                    else:\n",
      "                        new_object[name] = value\n",
      "            accession = post('/biosamples/', new_object, dry_run)\n",
      "            print('row {} ({}) -> {}'.format(i, new_object.get('aliases', None), accession))\n",
      "post_biosamples(book_name, True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def post_experiments(book_name, dry_run=False):\n",
      "    book = gcat.get_file(book_name, fmt='pandas_excel')\n",
      "    experiments = book.parse('Experiments', header=0)    \n",
      "    for i in experiments.index:\n",
      "        row = experiments.ix[i]\n",
      "        columns = experiments.columns\n",
      "        accession = row['experiment_accession']\n",
      "        new_object = {}\n",
      "        for k in columns:\n",
      "            if pandas.notnull(row[k]):\n",
      "                name, value = parse_sheet_cell(k, row[k])\n",
      "                if name is None:\n",
      "                    continue\n",
      "                new_object[name] = value\n",
      "        if pandas.isnull(accession):\n",
      "            accession = post('/experiments/', new_object, dry_run)\n",
      "        print('row {} ({}) -> {}'.format(i, new_object.get('description', None), accession))\n",
      "post_experiments(book_name, True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "row 0 (GM12878_30B) -> skip Old ENCSR000AIZ\n",
        "row 1 (PGP human bipolar neurons-total-RNA-13pgs) -> skip Old ENCSR225BBK\n",
        "row 2 (Total RNA-seq on 10ng of embryonic 11.5 day mouse forebrain) -> ENCSR160IIN\n",
        "row 3 (Total RNA-seq on 10ng of embryonic 11.5 day mouse liver) -> ENCSR284AMY\n"
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def update_model_with_flowcells(model):\n",
      "    query = RDF.SPARQLQuery(\"\"\"PREFIX htsw: <http://jumpgate.caltech.edu/wiki/LibraryOntology#>\n",
      "    select ?flowcell\n",
      "    where {\n",
      "       ?lane a htsw:IlluminaLane ;\n",
      "             htsw:flowcell ?flowcell .\n",
      "    }\"\"\")\n",
      "    for row in query.execute(model):\n",
      "        load_into_model(model, 'rdfa', str(row['flowcell']))\n",
      "\n",
      "def add_flowcell_details(lib_id, new_object):\n",
      "    model = get_model()\n",
      "    load_into_model(model, 'rdfa', 'https://jumpgate.caltech.edu/library/{}/'.format(lib_id))\n",
      "    update_model_with_flowcells(model)\n",
      "    query = RDF.SPARQLQuery(\"\"\"PREFIX htsw: <http://jumpgate.caltech.edu/wiki/LibraryOntology#>\n",
      "    select ?library ?flowcell_id ?lane_number ?barcode ?sequencer\n",
      "    where {{\n",
      "      ?library a htsw:Library ;\n",
      "               htsw:has_lane ?lane ;\n",
      "               htsw:multiplex_index ?barcode .\n",
      "      ?lane a htsw:IlluminaLane ;\n",
      "            htsw:flowcell ?flowcell ;\n",
      "            htsw:lane_number ?lane_number .\n",
      "      ?flowcell htsw:sequenced_by ?sequencer ;\n",
      "                htsw:flowcell_id ?flowcell_id .\n",
      "    }}\n",
      "    \"\"\")\n",
      "    flowcell_details = []\n",
      "    for row in query.execute(model):\n",
      "        run = {\n",
      "           'machine': str(row['sequencer']),\n",
      "           'flowcell': str(row['flowcell_id']),\n",
      "           'lane': str(row['lane_number']),\n",
      "           'barcode': str(row['barcode']).split(':')[-1]\n",
      "        }\n",
      "        flowcell_details.append(run)\n",
      "    new_object['flowcell_details'] = flowcell_details\n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def post_libraries(book_name, dry_run=False):\n",
      "    book = gcat.get_file(book_name, fmt='pandas_excel')\n",
      "    libraries = book.parse('Libraries', header = 0)\n",
      "    for i in libraries.index:\n",
      "        row = libraries.ix[i]\n",
      "        columns = libraries.columns\n",
      "        \n",
      "        lib_object = {}\n",
      "        rep_object = {}\n",
      "        for k in columns:\n",
      "            if pandas.notnull(row[k]):\n",
      "                name, value = parse_sheet_cell(k, row[k])\n",
      "                if name is None:\n",
      "                    continue\n",
      "                if name in ('experiment_accession', 'biological_replicate_number', 'technical_replicate_number', \n",
      "                            'replicate_uuid'):\n",
      "                    if name == 'experiment_accession':\n",
      "                        name = 'experiment'\n",
      "                    rep_object[name] = value\n",
      "                else:\n",
      "                    lib_object[name] = value\n",
      "        lib_accession = lib_object.get('library_accession', None)\n",
      "        if lib_accession is None:\n",
      "            lib_accession = post('/libraries/', lib_object, dry_run)\n",
      "        lib_id = None\n",
      "        if 'aliases' in lib_object:\n",
      "            lib_alias = lib_object['aliases'][0]\n",
      "            lib_id = lib_alias.replace('barbara-wold:','')\n",
      "        print('row {} lib ({}) -> {}'.format(i, lib_id, lib_accession))\n",
      "\n",
      "        if 'replicate_uuid' not in rep_object and \\\n",
      "           lib_accession is not None and \\\n",
      "           not lib_accession.startswith('skip'):\n",
      "            rep_object['library'] = lib_alias\n",
      "            add_flowcell_details(lib_id, rep_object)\n",
      "            #display.display_pretty(rep_object)\n",
      "            rep_accession = post('/replicates/', rep_object, dry_run)\n",
      "            print('row {} rep( {}-{}-{}) -> {}'.format(i, \n",
      "                                             rep_object['experiment'], \n",
      "                                             rep_object['biological_replicate_number'],\n",
      "                                             rep_object['technical_replicate_number'],\n",
      "                                             rep_accession))\n",
      "post_libraries(book_name, True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "row 0 lib (12517) -> skip example ENCLB482LEX\n",
        "row 1 lib (12518) -> skip example ENCLB898RJN\n",
        "row 2 lib (13413) -> skip\n",
        "row 3 lib (13414) -> skip\n",
        "row 4 lib (13415) -> skip\n",
        "row 5 lib (13416) -> skip\n",
        "row 6 lib (13417) -> skip\n",
        "row 7 lib (13418) -> skip\n",
        "row 8 lib (13419) -> skip\n",
        "row 9 lib (13420) -> skip\n",
        "row 10 lib (13421) -> skip\n",
        "row 11 lib (13422) -> skip\n",
        "row 12 lib (13423) -> skip\n",
        "row 13 lib (13424) -> skip\n",
        "row 14 lib (13425) -> skip\n",
        "row 15 lib (13426) -> skip\n",
        "row 16 lib (13427) -> skip\n",
        "row 17 lib (13428) -> skip\n",
        "row 18 lib (13429) -> skip\n",
        "row 19 lib (13430) -> skip\n",
        "row 20 lib (13431) -> skip\n",
        "row 21 lib (13432) -> skip\n",
        "row 22 lib (13433) -> skip\n",
        "row 23 lib (13434) -> skip\n",
        "row 24 lib (13435) -> skip\n",
        "row 25 lib (13436) -> skip\n",
        "row 26 lib (13437) -> skip\n",
        "row 27 lib (13438) -> skip\n",
        "row 28 lib (13439) -> skip\n",
        "row 29 lib (13440) -> skip\n",
        "row 30 lib (13443) -> skip\n",
        "row 31 lib (13444) -> skip\n",
        "row 32 lib (13445) -> skip\n",
        "row 33 lib (13446) -> skip\n",
        "row 34 lib (13447) -> skip\n",
        "row 35 lib (13448) -> skip\n",
        "row 36 lib (15007) -> ENCLB835LVO\n",
        "row 37 lib (15008) -> ENCLB471QMM\n",
        "row 38 lib (15009) -> ENCLB659UFY\n",
        "row 39 lib (15010) -> ENCLB426JKF\n",
        "row 40 lib (15011) -> ENCLB454ZUS\n",
        "row 41 lib (15012) -> ENCLB061TDP\n",
        "row 42 lib (15013) -> ENCLB331VAA\n",
        "row 43 lib (15014) -> ENCLB761PCA\n",
        "row 44 lib (15015) -> ENCLB601XLL\n",
        "row 45 lib (15016) -> ENCLB347FRI\n",
        "row 46 lib (15017) -> ENCLB055KYV\n",
        "row 47 lib (15018) -> ENCLB357KYA\n",
        "row 48 lib (15019) -> ENCLB495FUZ\n",
        "row 49 lib (15020) -> ENCLB979DZW\n",
        "row 50 lib (15021) -> ENCLB162EJJ\n",
        "row 51 lib (15022) -> ENCLB627LDZ\n",
        "row 52 lib (None) -> skip\n",
        "row 53 lib (None) -> skip\n",
        "row 54 lib (None) -> skip\n",
        "row 55 lib (None) -> skip\n",
        "row 56 lib (None) -> skip\n",
        "row 57 lib (None) -> skip\n",
        "row 58 lib (None) -> skip\n",
        "row 59 lib (None) -> skip\n",
        "row 60 lib (None) -> skip\n",
        "row 61 lib (None) -> skip\n",
        "row 62 lib (None) -> skip\n"
       ]
      }
     ],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import paramiko"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 15
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def make_manifest(book_name, map_name):\n",
      "    book = gcat.get_file(book_name, fmt='pandas_excel')\n",
      "    libraries = book.parse('Libraries', header = 0)\n",
      "    pongo = paramiko.SSHClient()\n",
      "    pongo.set_missing_host_key_policy(paramiko.WarningPolicy())\n",
      "    pongo.connect('thornvalley.caltech.edu', username='diane')\n",
      "    sftp = pongo.open_sftp()\n",
      "    sftp.chdir('proj/submission/encode-y3q1')\n",
      "\n",
      "    for line in sftp.open(map_name):\n",
      "        library_id, dirname = line.split()\n",
      "        row = libraries[libraries['aliases:array'] == 'barbara-wold:{}'.format(library_id)]\n",
      "        for filename in sftp.listdir(dirname):\n",
      "            if filename.endswith('fastq.gz'):\n",
      "                relative_path = os.path.join(dirname, filename)\n",
      "            metadata = {\n",
      "                'dataset': list(row['experiment_accession'])[0],\n",
      "                'submitted_file_name': relative_path,\n",
      "                'file_format': 'fastq',\n",
      "                'file_size': sftp.stat(relative_path).st_size,\n",
      "                #'md5sum': hashlib.make_md5sum(relative_path),\n",
      "                'output_type': 'raw data',\n",
      "                'replicate': list(row['replicate_uuid'])[0],\n",
      "                'lab': list(row['lab'])[0],\n",
      "                'award': list(row['award'])[0],\n",
      "            }\n",
      "            #display.display_pretty(metadata)\n",
      "            print('<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#'+dirname+'>')\n",
      "            print('  encode3:dataset \"{}\" ;'.format(metadata['dataset']))\n",
      "            print('  encode3:replicate \"{}\" ;'.format(metadata['replicate']))\n",
      "            print('  encode3:lab \"{}\" ;'.format(metadata['lab']))\n",
      "            print('  encode3:award \"{}\" .'.format(metadata['award']))\n",
      "\n",
      "            print('<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/'+relative_path+'>')\n",
      "            print('  encode3:output_type \"raw data\" ;')\n",
      "            print('  encode3:file_format \"fastq\" ;')\n",
      "            print('  encode3:file_size {} .'.format(metadata['file_size']))\n",
      "\n",
      "    print('generated')\n",
      "\n",
      "make_manifest(book_name, 'enhancer.txt')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15007-e11.5-forebrain1_10ngs>\n",
        "  encode3:dataset \"ENCSR160IIN\" ;\n",
        "  encode3:replicate \"62b52dad-8690-4e49-9247-5375f70794bf\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15007-e11.5-forebrain1_10ngs/15007_HAM6KADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2694720096 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15007-e11.5-forebrain1_10ngs>\n",
        "  encode3:dataset \"ENCSR160IIN\" ;\n",
        "  encode3:replicate \"62b52dad-8690-4e49-9247-5375f70794bf\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15007-e11.5-forebrain1_10ngs/15007_HAM6KADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2694720096 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15007-e11.5-forebrain1_10ngs>\n",
        "  encode3:dataset \"ENCSR160IIN\" ;\n",
        "  encode3:replicate \"62b52dad-8690-4e49-9247-5375f70794bf\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15007-e11.5-forebrain1_10ngs/15007_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2707628984 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15007-e11.5-forebrain1_10ngs>\n",
        "  encode3:dataset \"ENCSR160IIN\" ;\n",
        "  encode3:replicate \"62b52dad-8690-4e49-9247-5375f70794bf\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15007-e11.5-forebrain1_10ngs/15007_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2707628984 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15008-e11.5-forebrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR160IIN\" ;\n",
        "  encode3:replicate \"c96cad83-ae72-4ab6-9054-75865cef0a13\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15007-e11.5-forebrain1_10ngs/15007_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2707628984 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15008-e11.5-forebrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR160IIN\" ;\n",
        "  encode3:replicate \"c96cad83-ae72-4ab6-9054-75865cef0a13\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15008-e11.5-forebrain2_10ngs/15008_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2011317989 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15008-e11.5-forebrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR160IIN\" ;\n",
        "  encode3:replicate \"c96cad83-ae72-4ab6-9054-75865cef0a13\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15008-e11.5-forebrain2_10ngs/15008_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2011317989 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15008-e11.5-forebrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR160IIN\" ;\n",
        "  encode3:replicate \"c96cad83-ae72-4ab6-9054-75865cef0a13\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15008-e11.5-forebrain2_10ngs/15008_HAM6KADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1995638840 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15009-e11.5-midbrain1_10ngs>\n",
        "  encode3:dataset \"ENCSR307BCA\" ;\n",
        "  encode3:replicate \"d6e186fa-099d-4100-9f53-fb16fd364a2f\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15008-e11.5-forebrain2_10ngs/15008_HAM6KADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1995638840 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15009-e11.5-midbrain1_10ngs>\n",
        "  encode3:dataset \"ENCSR307BCA\" ;\n",
        "  encode3:replicate \"d6e186fa-099d-4100-9f53-fb16fd364a2f\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15009-e11.5-midbrain1_10ngs/15009_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2176561632 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15009-e11.5-midbrain1_10ngs>\n",
        "  encode3:dataset \"ENCSR307BCA\" ;\n",
        "  encode3:replicate \"d6e186fa-099d-4100-9f53-fb16fd364a2f\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15009-e11.5-midbrain1_10ngs/15009_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2176561632 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15009-e11.5-midbrain1_10ngs>\n",
        "  encode3:dataset \"ENCSR307BCA\" ;\n",
        "  encode3:replicate \"d6e186fa-099d-4100-9f53-fb16fd364a2f\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15009-e11.5-midbrain1_10ngs/15009_HAM6KADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2159438664 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15010-e11.5-midbrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR307BCA\" ;\n",
        "  encode3:replicate \"43b2386d-c1c4-4fb7-8c9d-5349ae4279aa\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15009-e11.5-midbrain1_10ngs/15009_HAM6KADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2159438664 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15010-e11.5-midbrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR307BCA\" ;\n",
        "  encode3:replicate \"43b2386d-c1c4-4fb7-8c9d-5349ae4279aa\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15010-e11.5-midbrain2_10ngs/15010_HAM6KADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2568766706 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15010-e11.5-midbrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR307BCA\" ;\n",
        "  encode3:replicate \"43b2386d-c1c4-4fb7-8c9d-5349ae4279aa\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15010-e11.5-midbrain2_10ngs/15010_HAM6KADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2568766706 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15010-e11.5-midbrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR307BCA\" ;\n",
        "  encode3:replicate \"43b2386d-c1c4-4fb7-8c9d-5349ae4279aa\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15010-e11.5-midbrain2_10ngs/15010_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2578097377 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15011-e11.5-hindbrain1_10ngs>"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "  encode3:dataset \"ENCSR760TOE\" ;\n",
        "  encode3:replicate \"f0d3e862-756b-4e20-a712-99bcd3dc37f4\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15010-e11.5-midbrain2_10ngs/15010_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2578097377 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15011-e11.5-hindbrain1_10ngs>\n",
        "  encode3:dataset \"ENCSR760TOE\" ;\n",
        "  encode3:replicate \"f0d3e862-756b-4e20-a712-99bcd3dc37f4\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15011-e11.5-hindbrain1_10ngs/15011_HAM6KADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2364449437 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15011-e11.5-hindbrain1_10ngs>\n",
        "  encode3:dataset \"ENCSR760TOE\" ;\n",
        "  encode3:replicate \"f0d3e862-756b-4e20-a712-99bcd3dc37f4\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15011-e11.5-hindbrain1_10ngs/15011_HAM6KADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2364449437 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15011-e11.5-hindbrain1_10ngs>\n",
        "  encode3:dataset \"ENCSR760TOE\" ;\n",
        "  encode3:replicate \"f0d3e862-756b-4e20-a712-99bcd3dc37f4\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15011-e11.5-hindbrain1_10ngs/15011_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2386257130 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15012-e11.5-hindbrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR760TOE\" ;\n",
        "  encode3:replicate \"9461c201-0667-40a4-adc8-e5d89e0afa08\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15012-e11.5-hindbrain2_10ngs/15012_HAM6KADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2361089850 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15012-e11.5-hindbrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR760TOE\" ;\n",
        "  encode3:replicate \"9461c201-0667-40a4-adc8-e5d89e0afa08\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15012-e11.5-hindbrain2_10ngs/15012_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2378962469 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15012-e11.5-hindbrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR760TOE\" ;\n",
        "  encode3:replicate \"9461c201-0667-40a4-adc8-e5d89e0afa08\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15012-e11.5-hindbrain2_10ngs/15012_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2378962469 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15012-e11.5-hindbrain2_10ngs>\n",
        "  encode3:dataset \"ENCSR760TOE\" ;\n",
        "  encode3:replicate \"9461c201-0667-40a4-adc8-e5d89e0afa08\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15012-e11.5-hindbrain2_10ngs/15012_HAM6KADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2378962469 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15013-e11.5_cranioface1_10ngs>\n",
        "  encode3:dataset \"ENCSR848HOX\" ;\n",
        "  encode3:replicate \"68aabdd2-3cde-4cf7-9eda-d60bca9a3d1d\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15013-e11.5_cranioface1_10ngs/15013_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1676246913 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15013-e11.5_cranioface1_10ngs>\n",
        "  encode3:dataset \"ENCSR848HOX\" ;\n",
        "  encode3:replicate \"68aabdd2-3cde-4cf7-9eda-d60bca9a3d1d\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15013-e11.5_cranioface1_10ngs/15013_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1676246913 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15013-e11.5_cranioface1_10ngs>\n",
        "  encode3:dataset \"ENCSR848HOX\" ;\n",
        "  encode3:replicate \"68aabdd2-3cde-4cf7-9eda-d60bca9a3d1d\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15013-e11.5_cranioface1_10ngs/15013_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1685048340 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15013-e11.5_cranioface1_10ngs>\n",
        "  encode3:dataset \"ENCSR848HOX\" ;\n",
        "  encode3:replicate \"68aabdd2-3cde-4cf7-9eda-d60bca9a3d1d\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15013-e11.5_cranioface1_10ngs/15013_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1685048340 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15014-e11.5_cranioface2_10ngs>\n",
        "  encode3:dataset \"ENCSR848HOX\" ;\n",
        "  encode3:replicate \"0518dba0-85b6-4bed-953d-44f68583af43\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15014-e11.5_cranioface2_10ngs/15014_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1728668596 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15014-e11.5_cranioface2_10ngs>\n",
        "  encode3:dataset \"ENCSR848HOX\" ;\n",
        "  encode3:replicate \"0518dba0-85b6-4bed-953d-44f68583af43\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15014-e11.5_cranioface2_10ngs/15014_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1728668596 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15014-e11.5_cranioface2_10ngs>\n",
        "  encode3:dataset \"ENCSR848HOX\" ;\n",
        "  encode3:replicate \"0518dba0-85b6-4bed-953d-44f68583af43\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15014-e11.5_cranioface2_10ngs/15014_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1728668596 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15014-e11.5_cranioface2_10ngs>\n",
        "  encode3:dataset \"ENCSR848HOX\" ;\n",
        "  encode3:replicate \"0518dba0-85b6-4bed-953d-44f68583af43\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15014-e11.5_cranioface2_10ngs/15014_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1736229441 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15015-e11.5_heart1_10ngs>\n",
        "  encode3:dataset \"ENCSR691OPQ\" ;\n",
        "  encode3:replicate \"bd583dac-77ab-4b0c-9f35-037ec609b933\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15014-e11.5_cranioface2_10ngs/15014_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1736229441 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15015-e11.5_heart1_10ngs>\n",
        "  encode3:dataset \"ENCSR691OPQ\" ;\n",
        "  encode3:replicate \"bd583dac-77ab-4b0c-9f35-037ec609b933\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15015-e11.5_heart1_10ngs/15015_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2222479571 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15015-e11.5_heart1_10ngs>\n",
        "  encode3:dataset \"ENCSR691OPQ\" ;\n",
        "  encode3:replicate \"bd583dac-77ab-4b0c-9f35-037ec609b933\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15015-e11.5_heart1_10ngs/15015_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2238450745 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15015-e11.5_heart1_10ngs>\n",
        "  encode3:dataset \"ENCSR691OPQ\" ;\n",
        "  encode3:replicate \"bd583dac-77ab-4b0c-9f35-037ec609b933\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15015-e11.5_heart1_10ngs/15015_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2238450745 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15016-e11.5_heart2_10ngs>"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "  encode3:dataset \"ENCSR691OPQ\" ;\n",
        "  encode3:replicate \"5708c8ba-54c8-44a1-aafd-1bc179341a1a\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15016-e11.5_heart2_10ngs/15016_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1554141058 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15016-e11.5_heart2_10ngs>\n",
        "  encode3:dataset \"ENCSR691OPQ\" ;\n",
        "  encode3:replicate \"5708c8ba-54c8-44a1-aafd-1bc179341a1a\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15016-e11.5_heart2_10ngs/15016_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1554141058 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15016-e11.5_heart2_10ngs>\n",
        "  encode3:dataset \"ENCSR691OPQ\" ;\n",
        "  encode3:replicate \"5708c8ba-54c8-44a1-aafd-1bc179341a1a\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15016-e11.5_heart2_10ngs/15016_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1557751919 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15016-e11.5_heart2_10ngs>\n",
        "  encode3:dataset \"ENCSR691OPQ\" ;\n",
        "  encode3:replicate \"5708c8ba-54c8-44a1-aafd-1bc179341a1a\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15016-e11.5_heart2_10ngs/15016_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1557751919 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15017-e11.5_liver1_10ngs>\n",
        "  encode3:dataset \"ENCSR284AMY\" ;\n",
        "  encode3:replicate \"02773643-83b5-4ecf-b903-f3b7cc3d4075\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15017-e11.5_liver1_10ngs/15017_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2004108008 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15017-e11.5_liver1_10ngs>\n",
        "  encode3:dataset \"ENCSR284AMY\" ;\n",
        "  encode3:replicate \"02773643-83b5-4ecf-b903-f3b7cc3d4075\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15017-e11.5_liver1_10ngs/15017_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2004108008 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15017-e11.5_liver1_10ngs>\n",
        "  encode3:dataset \"ENCSR284AMY\" ;\n",
        "  encode3:replicate \"02773643-83b5-4ecf-b903-f3b7cc3d4075\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15017-e11.5_liver1_10ngs/15017_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2004108008 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15017-e11.5_liver1_10ngs>\n",
        "  encode3:dataset \"ENCSR284AMY\" ;\n",
        "  encode3:replicate \"02773643-83b5-4ecf-b903-f3b7cc3d4075\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15017-e11.5_liver1_10ngs/15017_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 2015471401 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15018-e11.5_liver2_10ngs>\n",
        "  encode3:dataset \"ENCSR284AMY\" ;\n",
        "  encode3:replicate \"4be20e21-6cdb-4563-9786-985109589e10\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15018-e11.5_liver2_10ngs/15018_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1768660155 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15018-e11.5_liver2_10ngs>\n",
        "  encode3:dataset \"ENCSR284AMY\" ;\n",
        "  encode3:replicate \"4be20e21-6cdb-4563-9786-985109589e10\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15018-e11.5_liver2_10ngs/15018_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1768660155 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15018-e11.5_liver2_10ngs>\n",
        "  encode3:dataset \"ENCSR284AMY\" ;\n",
        "  encode3:replicate \"4be20e21-6cdb-4563-9786-985109589e10\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15018-e11.5_liver2_10ngs/15018_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1758955095 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15018-e11.5_liver2_10ngs>\n",
        "  encode3:dataset \"ENCSR284AMY\" ;\n",
        "  encode3:replicate \"4be20e21-6cdb-4563-9786-985109589e10\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15018-e11.5_liver2_10ngs/15018_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1758955095 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15021-e11.5_neuraltube1_10ngs>\n",
        "  encode3:dataset \"ENCSR337FYI\" ;\n",
        "  encode3:replicate \"9eda27a3-c132-4881-b4ef-35efbff0e273\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15021-e11.5_neuraltube1_10ngs/15021_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1561292493 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15021-e11.5_neuraltube1_10ngs>\n",
        "  encode3:dataset \"ENCSR337FYI\" ;\n",
        "  encode3:replicate \"9eda27a3-c132-4881-b4ef-35efbff0e273\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15021-e11.5_neuraltube1_10ngs/15021_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1561292493 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15021-e11.5_neuraltube1_10ngs>\n",
        "  encode3:dataset \"ENCSR337FYI\" ;\n",
        "  encode3:replicate \"9eda27a3-c132-4881-b4ef-35efbff0e273\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15021-e11.5_neuraltube1_10ngs/15021_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1561292493 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15021-e11.5_neuraltube1_10ngs>\n",
        "  encode3:dataset \"ENCSR337FYI\" ;\n",
        "  encode3:replicate \"9eda27a3-c132-4881-b4ef-35efbff0e273\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15021-e11.5_neuraltube1_10ngs/15021_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1551299302 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15022-e11.5_neuraltube2_10ngs>\n",
        "  encode3:dataset \"ENCSR337FYI\" ;\n",
        "  encode3:replicate \"aac3c45f-5748-4f4b-bdca-e012ed477453\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15021-e11.5_neuraltube1_10ngs/15021_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1551299302 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15022-e11.5_neuraltube2_10ngs>\n",
        "  encode3:dataset \"ENCSR337FYI\" ;\n",
        "  encode3:replicate \"aac3c45f-5748-4f4b-bdca-e012ed477453\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15021-e11.5_neuraltube1_10ngs/15021_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1551299302 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15022-e11.5_neuraltube2_10ngs>\n",
        "  encode3:dataset \"ENCSR337FYI\" ;\n",
        "  encode3:replicate \"aac3c45f-5748-4f4b-bdca-e012ed477453\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15022-e11.5_neuraltube2_10ngs/15022_HANW6ADXX_c116_l2.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1613237917 .\n",
        "<http://jumpgate.caltech.edu/wiki/SubmissionsLog/encode-y3q1#15022-e11.5_neuraltube2_10ngs>\n",
        "  encode3:dataset \"ENCSR337FYI\" ;\n",
        "  encode3:replicate \"aac3c45f-5748-4f4b-bdca-e012ed477453\" ;\n",
        "  encode3:lab \"barbara-wold\" ;\n",
        "  encode3:award \"U54HG006998\" .\n",
        "<file:///woldlab/castor/home/diane/proj/submission/encode-y3q1/15022-e11.5_neuraltube2_10ngs/15022_HANW6ADXX_c116_l1.fastq.gz>\n",
        "  encode3:output_type \"raw data\" ;\n",
        "  encode3:file_format \"fastq\" ;\n",
        "  encode3:file_size 1620653264 .\n",
        "generated\n"
       ]
      }
     ],
     "prompt_number": 29
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%debug"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "> \u001b[0;32m<ipython-input-24-32107639da6f>\u001b[0m(20)\u001b[0;36mmake_manifest\u001b[0;34m()\u001b[0m\n",
        "\u001b[0;32m     19 \u001b[0;31m                \u001b[0;34m'file_format'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'fastq'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0m\u001b[0;32m---> 20 \u001b[0;31m                \u001b[0;34m'file_size'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0msftp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrelative_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0m\u001b[0;32m     21 \u001b[0;31m                \u001b[0;31m#'md5sum': hashlib.make_md5sum(relative_path),\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0m\n"
       ]
      },
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "ipdb> p dir(sftp.stat(relative_path)\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "*** SyntaxError: SyntaxError('unexpected EOF while parsing', ('<string>', 1, 28, 'dir(sftp.stat(relative_path)'))\n"
       ]
      },
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "ipdb> p dir(sftp.stat(relative_path))\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "['FLAG_AMTIME', 'FLAG_EXTENDED', 'FLAG_PERMISSIONS', 'FLAG_SIZE', 'FLAG_UIDGID', '__class__', '__delattr__', '__dict__', '__doc__', '__format__', '__getattribute__', '__hash__', '__init__', '__module__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_debug_str', '_flags', '_from_msg', '_pack', '_rwx', '_unpack', 'asbytes', 'attr', 'from_stat', 'st_atime', 'st_gid', 'st_mode', 'st_mtime', 'st_size', 'st_uid']\n"
       ]
      },
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "ipdb> p sftp.stat(relative_path).st_size\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "2694720096\n"
       ]
      },
      {
       "name": "stdout",
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "ipdb> c\n"
       ]
      }
     ],
     "prompt_number": 25
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def add_document_aliases():\n",
      "    # Really I wish I could attach some aliases to the protocols\n",
      "    aliases = [#('f0cc5a7f-96a5-4970-9f46-317cc8e2d6a4', 'barbara-wold:mirvana_protocol.pdf'),\n",
      "               ('428a184d-7fa1-4599-9d8d-749c2eba7edd', 'barbara-wold:turbodnase_protocol.pdf'),\n",
      "               ('e84de88e-4c31-4b2a-b6ac-96feb764b2d0', 'barbara-wold:dynabeads_protocol.pdf'),\n",
      "               ('5f5193b0-0c61-40c6-a286-b1f6d6f0c0da', 'barbara-wold:Smart-Seq_Protocol.pdf'),\n",
      "               ]\n",
      "    for uuid, alias in aliases:\n",
      "        document = server.get_json(uuid)\n",
      "        if alias is None:\n",
      "            print('IDENTIFY: {} -> {}'.format(document['@id'], document['urls']))\n",
      "        elif alias not in document['aliases']:\n",
      "            print('UPDATE: {} -> {}'.format(document['@id'], document['aliases']))\n",
      "            document_aliases = document['aliases']\n",
      "            document_aliases.append(alias)\n",
      "            payload = {'aliases': document_aliases}\n",
      "            display.display_pretty(payload)\n",
      "            server.patch_json(document['@id'], payload)\n",
      "            \n",
      "#add_document_aliases()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "server.get_json('/profiles/file.json')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 30,
       "text": [
        "{u'$schema': u'http://json-schema.org/draft-04/schema#',\n",
        " u'additionalProperties': False,\n",
        " u'calculated_props': {u'href': {u'title': u'Download URL',\n",
        "   u'type': u'string'}},\n",
        " u'columns': {u'accession': {u'title': u'Accession', u'type': u'string'},\n",
        "  u'dataset': {u'title': u'Dataset', u'type': u'string'},\n",
        "  u'file_format': {u'title': u'File Format', u'type': u'string'},\n",
        "  u'md5sum': {u'title': u'MD5 Sum', u'type': u'string'},\n",
        "  u'output_type': {u'title': u'Output Type', u'type': u'string'}},\n",
        " u'dependencies': {u'file_format': {u'oneOf': [{u'properties': {u'file_format': {u'enum': [u'fastq']}},\n",
        "     u'required': [u'replicate']},\n",
        "    {u'not': {u'properties': {u'file_format': {u'enum': [u'fastq']}}}}]},\n",
        "  u'paired_end': {u'oneOf': [{u'properties': {u'paired_end': {u'enum': [u'2']}},\n",
        "     u'required': [u'paired_with']},\n",
        "    {u'properties': {u'paired_end': {u'enum': [u'1']}}}]},\n",
        "  u'paired_with': {u'properties': {u'file_format': {u'enum': [u'fastq',\n",
        "      u'csfasta',\n",
        "      u'csqual']},\n",
        "    u'paired_end': {u'enum': [u'2']}},\n",
        "   u'required': [u'paired_end']}},\n",
        " u'description': u'Schema for submitting metadata for a data file.',\n",
        " u'facets': {u'file_format': {u'title': u'File format', u'type': u'string'},\n",
        "  u'lab.title': {u'title': u'Lab', u'type': u'string'},\n",
        "  u'output_type': {u'title': u'Output type', u'type': u'string'},\n",
        "  u'status': {u'title': u'File status', u'type': u'string'}},\n",
        " u'file_format_file_extension': {u'2bit': u'.2bit',\n",
        "  u'CEL': u'.cel.gz',\n",
        "  u'bam': u'.bam',\n",
        "  u'bed': u'.bed.gz',\n",
        "  u'bedLogR': u'.bedLogR.bigBed  ',\n",
        "  u'bedMethyl': u'.bedMethyl.bigBed',\n",
        "  u'bedRnaElements': u'.bedRnaElements.bigBed',\n",
        "  u'bed_bedLogR': u'.bedLogR.gz',\n",
        "  u'bed_bedMethyl': u'.bedMethyl.gz',\n",
        "  u'bed_bedRnaElements': u'.bedRnaElements.gz',\n",
        "  u'bed_broadPeak': u'.broadPeak.gz',\n",
        "  u'bed_narrowPeak': u'.narrowPeak.gz',\n",
        "  u'bigBed': u'.bigBed',\n",
        "  u'bigWig': u'.bigWig',\n",
        "  u'broadPeak': u'.broadPeak.bigbed',\n",
        "  u'csfasta': u'.csfasta.gz',\n",
        "  u'csqual': u'.csqual.gz',\n",
        "  u'csv': u'.csv',\n",
        "  u'fasta': u'.fasta.gz',\n",
        "  u'fastq': u'.fastq.gz',\n",
        "  u'gff': u'.gff.gz',\n",
        "  u'gtf': u'.gtf.gz',\n",
        "  u'idat': u'.idat',\n",
        "  u'narrowPeak': u'.narrowPeak.bigBed',\n",
        "  u'rcc': u'.rcc',\n",
        "  u'tar': u'.tar.gz',\n",
        "  u'tsv': u'.tsv'},\n",
        " u'id': u'/profiles/file.json',\n",
        " u'identifyingProperties': [u'uuid', u'accession', u'aliases'],\n",
        " u'mixinProperties': [{u'$ref': u'mixins.json#/schema_version'},\n",
        "  {u'$ref': u'mixins.json#/uuid'},\n",
        "  {u'$ref': u'mixins.json#/accession'},\n",
        "  {u'$ref': u'mixins.json#/submitted'},\n",
        "  {u'$ref': u'mixins.json#/aliases'},\n",
        "  {u'$ref': u'mixins.json#/attribution'},\n",
        "  {u'$ref': u'mixins.json#/notes'}],\n",
        " u'properties': {u'accession': {u'accessionType': u'FF',\n",
        "   u'comment': u'Only admins are allowed to set or update this value.',\n",
        "   u'description': u'A unique identifier to be used to reference the object.',\n",
        "   u'format': u'accession',\n",
        "   u'permission': u'import_items',\n",
        "   u'serverDefault': u'accession',\n",
        "   u'title': u'Accession',\n",
        "   u'type': u'string'},\n",
        "  u'aliases': {u'default': [],\n",
        "   u'description': u'Lab specific identifiers to reference an object.',\n",
        "   u'items': {u'comment': u'Current convention is colon separated lab name and lab identifier. (e.g. john-doe:42).',\n",
        "    u'description': u'A lab specific identifier to reference an object.',\n",
        "    u'pattern': u'^\\\\S+:\\\\S+',\n",
        "    u'title': u'Lab alias',\n",
        "    u'type': u'string'},\n",
        "   u'title': u'Lab aliases',\n",
        "   u'type': u'array'},\n",
        "  u'alternate_accessions': {u'default': [],\n",
        "   u'description': u'Accessions previously assigned to objects that have been merged with this object.',\n",
        "   u'items': {u'comment': u'Only admins are allowed to set or update this value.',\n",
        "    u'description': u'An accession previously assigned to an object that has been merged with this object.',\n",
        "    u'format': u'accession',\n",
        "    u'title': u'Alternate Accession',\n",
        "    u'type': u'string'},\n",
        "   u'permission': u'import_items',\n",
        "   u'title': u'Alternate accessions',\n",
        "   u'type': u'array'},\n",
        "  u'assembly': {u'comment': u'Applies for mapped files (e.g. BAM, BED and BigWig).',\n",
        "   u'description': u'UCSC genome assembly that files were mapped to.',\n",
        "   u'enum': [u'mm9', u'hg19', u'mm10', u'ce10', u'dm3', u'dm6'],\n",
        "   u'title': u'Mapping assembly',\n",
        "   u'type': u'string'},\n",
        "  u'award': {u'comment': u'See award.json for list of available identifiers.',\n",
        "   u'description': u'Grant associated with the submission.',\n",
        "   u'linkTo': u'award',\n",
        "   u'title': u'Grant',\n",
        "   u'type': u'string'},\n",
        "  u'controlled_by': {u'description': u'The files that control this file. ',\n",
        "   u'items': {u'comment': u'See file.json for a list of available identifiers.',\n",
        "    u'linkTo': u'file',\n",
        "    u'type': u'string'},\n",
        "   u'title': u'Controlled by',\n",
        "   u'type': u'array'},\n",
        "  u'dataset': {u'comment': u'For experiments, see experiment.json for available identifiers. For datasets, see dataset.json for available identifiers.',\n",
        "   u'description': u'The experiment or dataset the file belongs to.',\n",
        "   u'linkTo': [u'experiment', u'dataset'],\n",
        "   u'title': u'Dataset',\n",
        "   u'type': u'string'},\n",
        "  u'date_created': {u'anyOf': [{u'format': u'date-time'},\n",
        "    {u'format': u'date'}],\n",
        "   u'comment': u'Do not submit, value is assigned by the server. The date the object is created.',\n",
        "   u'permission': u'import_items',\n",
        "   u'rdfs:subPropertyOf': u'dc:created',\n",
        "   u'serverDefault': u'now',\n",
        "   u'title': u'Date created',\n",
        "   u'type': u'string'},\n",
        "  u'derived_from': {u'description': u'The files participating as inputs into software to produce this output file.',\n",
        "   u'items': {u'comment': u'See file.json for a list of available identifiers.',\n",
        "    u'linkTo': u'file',\n",
        "    u'type': u'string'},\n",
        "   u'title': u'Derived from',\n",
        "   u'type': u'array'},\n",
        "  u'file_format': {u'oneOf': [{u'enum': [u'bam',\n",
        "      u'bed',\n",
        "      u'bedLogR',\n",
        "      u'bed_bedLogR',\n",
        "      u'bedMethyl',\n",
        "      u'bed_bedMethyl',\n",
        "      u'bigBed',\n",
        "      u'bigWig',\n",
        "      u'broadPeak',\n",
        "      u'bed_broadPeak',\n",
        "      u'fasta',\n",
        "      u'fastq',\n",
        "      u'gtf',\n",
        "      u'idat',\n",
        "      u'narrowPeak',\n",
        "      u'bed_narrowPeak',\n",
        "      u'rcc',\n",
        "      u'CEL',\n",
        "      u'tsv',\n",
        "      u'csv']},\n",
        "    {u'comment': u'Historical file formats, not valid for new submissions.',\n",
        "     u'enum': [u'tar', u'2bit', u'csfasta', u'csqual', u'bedRnaElements'],\n",
        "     u'permission': u'import_items'}],\n",
        "   u'title': u'File format',\n",
        "   u'type': u'string'},\n",
        "  u'file_size': {u'comment': u'File size is specified in bytes',\n",
        "   u'title': u'File size',\n",
        "   u'type': u'integer'},\n",
        "  u'flowcell_details': {u'default': [],\n",
        "   u'description': u'For high-throughput sequencing, the flowcells used for the sequencing of the replicate.',\n",
        "   u'items': {u'additionalProperties': False,\n",
        "    u'properties': {u'barcode': {u'title': u'Barcode', u'type': u'string'},\n",
        "     u'flowcell': {u'title': u'Flowcell ID', u'type': u'string'},\n",
        "     u'lane': {u'title': u'Lane', u'type': u'string'},\n",
        "     u'machine': {u'description': u'The lab specific name of the machine used.',\n",
        "      u'title': u'Machine Name',\n",
        "      u'type': u'string'}},\n",
        "    u'title': u'Flowcell details',\n",
        "    u'type': u'object'},\n",
        "   u'title': u'Flowcells',\n",
        "   u'type': u'array'},\n",
        "  u'lab': {u'comment': u'See lab.json for list of available identifiers.',\n",
        "   u'description': u'Lab associated with the submission.',\n",
        "   u'linkSubmitsFor': True,\n",
        "   u'linkTo': u'lab',\n",
        "   u'title': u'Lab',\n",
        "   u'type': u'string'},\n",
        "  u'md5sum': {u'format': u'hex', u'title': u'MD5sum', u'type': u'string'},\n",
        "  u'notes': {u'description': u'Additional information.',\n",
        "   u'permission': u'import_items',\n",
        "   u'title': u'Notes',\n",
        "   u'type': u'string'},\n",
        "  u'output_type': {u'description': u\"A description of the file's purpose or contents.\",\n",
        "   u'oneOf': [{u'enum': [u'CNV',\n",
        "      u'HMM',\n",
        "      u'alignments',\n",
        "      u'clusters',\n",
        "      u'contigs',\n",
        "      u'enhancers',\n",
        "      u'enrichment',\n",
        "      u'hotspots',\n",
        "      u'idat green file',\n",
        "      u'idat red file',\n",
        "      u'interactions',\n",
        "      u'library_fraction',\n",
        "      u'multi-read minus signal',\n",
        "      u'multi-read plus signal',\n",
        "      u'multi-read signal',\n",
        "      u'peaks',\n",
        "      u'plus signal',\n",
        "      u'raw data',\n",
        "      u'raw minus signal',\n",
        "      u'raw plus signal',\n",
        "      u'raw signal',\n",
        "      u'reads',\n",
        "      u'rejected reads',\n",
        "      u'signal',\n",
        "      u'sites',\n",
        "      u'spike-ins',\n",
        "      u'splice junctions',\n",
        "      u'unique minus signal',\n",
        "      u'unique plus signal',\n",
        "      u'unique signal']},\n",
        "    {u'comment': u'Historical output types, not valid for new submissions.',\n",
        "     u'enum': [u'Alignability',\n",
        "      u'Base_Overlap_Signal',\n",
        "      u'enhancers_forebrain',\n",
        "      u'enhancers_heart',\n",
        "      u'enhancers_wholebrain',\n",
        "      u'Excludable',\n",
        "      u'ExonsDeNovo',\n",
        "      u'ExonsEnsV65IAcuff',\n",
        "      u'ExonsGencV10',\n",
        "      u'ExonsGencV3c',\n",
        "      u'ExonsGencV7',\n",
        "      u'FiltTransfrags',\n",
        "      u'GeneDeNovo',\n",
        "      u'GeneEnsV65IAcuff',\n",
        "      u'GeneGencV10',\n",
        "      u'GeneGencV3c',\n",
        "      u'GeneGencV7',\n",
        "      u'Junctions',\n",
        "      u'Matrix',\n",
        "      u'minus signal',\n",
        "      u'mPepMapGcFt',\n",
        "      u'mPepMapGcUnFt',\n",
        "      u'PctSignal',\n",
        "      u'pepMapGcFt',\n",
        "      u'pepMapGcUnFt',\n",
        "      u'Primer',\n",
        "      u'PrimerPeaks',\n",
        "      u'RbpAssocRna',\n",
        "      u'SumSignal',\n",
        "      u'TranscriptDeNovo',\n",
        "      u'TranscriptEnsV65IAcuff',\n",
        "      u'TranscriptGencV10',\n",
        "      u'TranscriptGencV3c',\n",
        "      u'TranscriptGencV7',\n",
        "      u'Transfrags',\n",
        "      u'TssGencV3c',\n",
        "      u'TssGencV7',\n",
        "      u'TssHmm',\n",
        "      u'UniformlyProcessedPeakCalls',\n",
        "      u'Uniqueness',\n",
        "      u'Validation',\n",
        "      u'Valleys',\n",
        "      u'WaveSignal'],\n",
        "     u'permission': u'import_items'}],\n",
        "   u'title': u'Output type',\n",
        "   u'type': u'string'},\n",
        "  u'paired_end': {u'description': u'Which pair the file belongs to (if paired end library)',\n",
        "   u'enum': [u'1', u'2'],\n",
        "   u'title': u'Paired End Identifier',\n",
        "   u'type': u'string'},\n",
        "  u'paired_with': {u'comment': u'See file.json for a list of available identifiers.',\n",
        "   u'description': u'The file that corresponds with this file.',\n",
        "   u'linkTo': u'file',\n",
        "   u'title': u'File pairing',\n",
        "   u'type': u'string'},\n",
        "  u'platform': {u'comment': u'See platform.json for available identifiers.',\n",
        "   u'description': u'The measurement device used to collect data.',\n",
        "   u'linkTo': u'platform',\n",
        "   u'title': u'Platform',\n",
        "   u'type': u'string'},\n",
        "  u'replicate': {u'comment': u'See replicate.json for available identifiers.',\n",
        "   u'description': u'The experimental replicate designation for the file.',\n",
        "   u'linkTo': u'replicate',\n",
        "   u'title': u'Replicate',\n",
        "   u'type': u'string'},\n",
        "  u'schema_version': {u'comment': u'Do not submit, value is assigned by the server. The version of the JSON schema that the server uses to validate the object. Schema version indicates generation of schema used to save version to to enable upgrade steps to work. Individual schemas should set the default.',\n",
        "   u'default': u'4',\n",
        "   u'pattern': u'^\\\\d+(\\\\.\\\\d+)*$',\n",
        "   u'requestMethod': [],\n",
        "   u'type': u'string'},\n",
        "  u'status': {u'default': u'uploading',\n",
        "   u'enum': [u'uploading',\n",
        "    u'uploaded',\n",
        "    u'upload failed',\n",
        "    u'format check failed',\n",
        "    u'in progress',\n",
        "    u'deleted',\n",
        "    u'replaced',\n",
        "    u'revoked',\n",
        "    u'released'],\n",
        "   u'title': u'Status',\n",
        "   u'type': u'string'},\n",
        "  u'submitted_by': {u'comment': u'Do not submit, value is assigned by the server. The user that created the object.',\n",
        "   u'linkTo': u'user',\n",
        "   u'permission': u'import_items',\n",
        "   u'rdfs:subPropertyOf': u'dc:creator',\n",
        "   u'serverDefault': u'userid',\n",
        "   u'title': u'Submitted by',\n",
        "   u'type': u'string'},\n",
        "  u'submitted_file_name': {u'description': u'The local file name used at time of submission.',\n",
        "   u'title': u'Original file name',\n",
        "   u'type': u'string'},\n",
        "  u'supercedes': {u'description': u'The files that this file replaces.',\n",
        "   u'items': {u'comment': u'See file.json for a list of available identifiers.',\n",
        "    u'linkTo': u'file',\n",
        "    u'type': u'string'},\n",
        "   u'title': u'Supercedes',\n",
        "   u'type': u'array'},\n",
        "  u'uuid': {u'format': u'uuid',\n",
        "   u'permission': u'import_items',\n",
        "   u'requestMethod': u'POST',\n",
        "   u'serverDefault': u'uuid4',\n",
        "   u'title': u'UUID',\n",
        "   u'type': u'string'}},\n",
        " u'required': [u'dataset',\n",
        "  u'file_format',\n",
        "  u'md5sum',\n",
        "  u'output_type',\n",
        "  u'award',\n",
        "  u'lab'],\n",
        " u'title': u'Data file',\n",
        " u'type': u'object'}"
       ]
      }
     ],
     "prompt_number": 30
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}